### ISSUE TITLE
New JSON parser in runtime

### ISSUE THREAD

**Issue by troydai** [aspnet/Announcements#24]
We implemented a new JSON parser in `Microsoft.Framework.Runtime` and removed the dependency on `Newtonsoft.Json`. 

Using `Newtonsoft.Json` in the runtime prevents customers from using any version other than the one used in the runtime. This change will allow customers to use any version of `Newtonsoft.Json` in their applications.

The new JSON parser is customized for internal use. It is used to parse only `project.json`, `project.lock.json` and `global.json`. 

The new JSON parser doesn't support following syntax which was previously supported:
1. Comments in the form of `// ...` and `/* ... */`
2. Trailing commas after the last member of objects and arrays

Other components, such as `Microsoft.Framework.PackageManager`, still use `Newtonsoft.Json` to parse JSON files.

**Comment by aspnetde** [comment]
Are there any plans to support comments (again)? I think that would be really helpful.

**Comment by LordZoltan** [comment]
Yeah while I agree with the change to remove the dependency on NewtonSoft at this low-level, given the importance of these json files, it's perhaps a little naive to assume that they won't need to be annotated with comments for the benefit of humans...
On 12 May 2015 22:05, Thomas Bandt <notifications@github.com> wrote:Are there any plans to support comments (again)? I think that would be really helpful.

**Comment by NickCraver** [comment]
The lack of comment support is disappointing to say the least. Every developer I know following this was *very* excited about something with weight finally pushing for JSON comment support. This is really needed across the board - and I think Microsoft has a real opportunity to push adoption here with excellent reasons for doing so.

Every puppet configuration, hiera data file, or settings config we have here at Stack Overflow would benefit from comment support. I really feel you can and should push this. Older tools needing to read such files can run exclude comments via a regex before parsing if needed in the interim.

Please reconsider this. Comments are *needed* here, and in many, many other similar scenarios. This is an excellent use case to actually advance the JSON spec. Look at it this way: is there anyone who things JSON *shouldn't* support comments?

**Comment by stephen-riley** [comment]
Re: @NickCraver 
> is there anyone who things JSON shouldn't support comments?

No.  :-)

Seriously, let's make this happen.

**Comment by Eilon** [comment]
Hi folks, we're not opposed to adding support for comments - we completely agree they are useful. However, we wanted to start out with a more "pure" solution to reading JSON by following the JSON spec as closely as possible. We currently chose not to implement optional aspects of the spec.

We will certainly revisit support for comments and trailing commas.

**Comment by mikeobrien** [comment]
![no sir i dont like it](https://cloud.githubusercontent.com/assets/187817/7599011/b1373586-f8cc-11e4-9396-7a9bed5c5e82.gif)

**Comment by vcsjones** [comment]
I'll throw my vote in for supporting comments (trailing commas I can live without). I already have comments in my project.json file, and I would like to keep them. While "optional" I don't think customers will think of it as such. It's quite an omission.

**Comment by glen-84** [comment]
> Microsoft should push for comments to be included in the JSON spec. The reasons for it not being included are weak.

And ...

> The moment a parser expands the feature set, the document can no longer be parsed by other parsers. It's not at all intuitive that you can comment in project.json but not in bower.json or package.json. This is why it should be part of the spec once again, and why I thought it would be interesting to see how the spec writers would respond to such a request from a large corporation.
    
See https://github.com/aspnet/Tooling/issues/13#issuecomment-98988817.

@Eilon Is this something that you could discuss with the spec writers?

**Comment by stephen-riley** [comment]
@Eilon , totally understand wanting to conform to the spec.  However, the grammar for JSON with comments is a strict superset of the official JSON grammar, yes?  In that case, what's the harm?

Addressing the link that @glen-84 sent out (good stuff there), we're not asking for VS2015 files to come standard with comments--by all means, leave comments out of the JSON documents that you want to be parseable by standard parsers.  We just want the option of including them ourselves in our own files, totally taking on all risk and blame if our files break other parsers.  "With great power..." and all that.

**Comment by jbogard** [comment]
Totally curious, what are you commenting in your project.json file?

**Comment by GuardRex** [comment]
Can we have an optional `comment` element for each section of the `project.json` schema? At least we would get one line per section ... and it wouldn't break the spec.

**Comment by glen-84** [comment]
@GuardRex :-1: That's fugly.

**Comment by Eilon** [comment]
The problem with comments in JSON files is interoperability with other systems that might read them. For example, even GitHub doesn't like comments in JSON files and they show up like this:

![image](https://cloud.githubusercontent.com/assets/202643/7599340/06b22c8a-f8b6-11e4-9f8e-57eacb42be80.png)

That's at the very least ugly, but in potentially worse cases it could mean that some editors/parsers would refuse to parse the file.

**Comment by glen-84** [comment]
@jbogard,

It could be anything from temporarily removing a dependency, to breaking dependencies into sections, describing what commands do, adding TODOs, etc.

**Comment by GuardRex** [comment]
@glen-84 Hence the word "optional" in my suggestion. I'm just saying ... if the spec is so all-fire important, then include an optional element in the schema and have the parser skip it. What's the big deal?

**Comment by kowsheek** [comment]
I've used it to remove temp dependencies as well.

But, as far as JSON goes, it's not technically supposed to have comments in there, so I understand the reasoning. But, we need to find a balance between usage and standard.

There could be a section for "not used" dependencies, similar to `exclude`.

**Comment by danielmarbach** [comment]
In the age of DVCS. Why the hell are we still commenting out? You can branch, stash… as many project.json file changes as you can dream of.

**Comment by glen-84** [comment]
@GuardRex It just seems like a hack to me.

**Comment by GuardRex** [comment]
@glen-84 It is ... just thinking how it keeps JSON spec cool and gives one line that you can put anything you want into. @danielmarbach I was thinking more along the lines of notes rather than commenting things out.

**Comment by CoreyKaylor** [comment]
Could care less about comments in JSON myself. Easy enough to get back deleted code from a previous commit. Annotate gives context as to why it's there. Odd that I never even considered commenting my *.csproj files either.

**Comment by glen-84** [comment]
@kowsheek, that's just one use case. I would not extend the schema for something like that.

**Comment by glen-84** [comment]
@danielmarbach, You seriously branch/stash and revert each time you comment in a C# file? o.O

**Comment by kowsheek** [comment]
@glen-84 true.

@danielmarbach Naah, sometimes you're in the middle of something and it comes in handy to comment out packages.

**Comment by NickCraver** [comment]
@danielmarbach Commenting out is not used just for disabling code & components. Think about the default `web.config` in vCurrent for example: "here's how you would do X". I would be using this in all my example configuration files for projects of pieces you can enable/change/etc. if JSON supported it already. DVCS is really an orthogonal concern to the many use cases here.

**Comment by danielmarbach** [comment]
Stage, stash, done

**Comment by GuardRex** [comment]
What's the current thinking on having normal comments `/* */` and/or `//` and using Gulp to strip them out prior to `project.json` being parsed? I see a [gulp-strip-json-comments](https://www.npmjs.com/package/gulp-strip-json-comments/) plugin, but I don't know how to fire it off prior to `project.json` getting parsed.  Can it be done?

**Comment by davidfowl** [comment]
No

**Comment by mgravell** [comment]
@Daniel by that logic: we shouldn't have comments in code either :) it isn't about deleting stuff; it is about **commenting** - why something is what it is, why it can't be something else - something unobvious that the reader should know

**Comment by GuardRex** [comment]
Rats! :rat: :rat: :rat:

**Comment by davidfowl** [comment]
Lets just all go force the JSON spec to support comments :smile: Bike shedding on this github announcement is pretty useless. I think we all understand people want comments.

**Comment by glen-84** [comment]
@davidfowl :+1: ... but like I said, a request/proposal from a large corporation would carry a lot more weight than one from a lone developer.

**Comment by mj1856** [comment]
If comments *were* to be supported, how would they play nice with tooling?  I would think they could be easily mutilated or lost.

**Comment by kowsheek** [comment]
With @davidfowl for this one. JSON specs don't allow comments, since JSON is built for data-interchange. I'm sure we won't miss it as much as all these comments here.

**Comment by glen-84** [comment]
Or you could just switch to [YAML](http://yaml.org/). :grin:

**Comment by davidfowl** [comment]
Or TOML.

**Comment by glen-84** [comment]
That looks like INI, lol.

**Comment by GuardRex** [comment]
Or [JSON5](http://json5.org/).

**Comment by NickCraver** [comment]
I'm not sure if JSON5 is a joke suggestion or not, but it actually fits the bill for many of the problems. With the differing file extension it also removes confusion of what should and shouldn't work with other parsers. I'm not saying it could be used outright, but adopting it or similar as a push to the spec (with Microsoft's weight) would be an excellent start. It doesn't really solve the tangentially related date issue, but it solves all the others in-line with all the other ECMAScript updates.

**Comment by atrauzzi** [comment]
Yeah, new/different format is warranted here.

Please do not switch to yaml.

**Comment by glen-84** [comment]
@GuardRex,

That actually looks perfect!

Two problems:

1. It's not as common as JSON, so there are less tools available. This is not a big deal IMHO.
2. It might be too late to make such a change to DNX.

Still, I'd +1 using a format like that.

**Comment by GuardRex** [comment]
I mentioned JSON5 because someone opened it for voting over at [Uservoice](https://aspnet.uservoice.com/forums/252111-asp-net-vnext/suggestions/7209422-use-json5-for-project-json).

**Comment by vbfox** [comment]
Please just not keep the ".json" file extension if you plan to make it something else than the specification (Like including comments before having them added to the spec) it'll be painful.

**Comment by mariusschulz** [comment]
@vbfox Full ack, that's what I was thinking, too. If the configuration is not pure (read: standardized) JSON, the file extension shouldn't be `.json`; otherwise, it'll just confuse editors and cause interop problems.

**Comment by JamesNK** [comment]
Douglas Crockford's thoughts on comments: https://plus.google.com/+DouglasCrockfordEsq/posts/RK8qyGVaGSr

**Comment by fekberg** [comment]
Could we add support for regions in json while we're at it? :shipit:

**Comment by atrauzzi** [comment]
Typescript?

**Comment by nulltoken** [comment]
Or, maybe should we take a step back and seriously ponder whether now wouldn't be a great time to re-introduce xml? :trollface:

**Comment by adamralph** [comment]
JSON is ostensibly a serialization format. This begs the question, why does a serialization format need comments?

I agree with those comments above which allude to stepping back and considering what JSON is, what the use case is and ultimately what the right solution is, before trying to potentially shoehorn something inappropriate into the JSON standard.

**Comment by blobor** [comment]
@troydai
If this change affects only beta releases, it's ok for me.
But do you planing add json comments support on the RC release?

**Comment by djanosik** [comment]
Oh. I think this debate is completely pointless. All mentioned files (`project.json`, `project.lock.json`, `global.json`) are very simple. There is no need to have comments there. You can still use comments in `config.json`, right?

**Comment by adamralph** [comment]
If config.json is JSON then the same argument applies.

**Comment by tugberkugurlu** [comment]
Just to state my point: you 99.9% of the time don't need comments on these files: project.json, project.lock.json, global.json. Make use of git blame to see why the changes actually happened.

@adamralph Microsoft.Framework.ConfigurationModel.Json project [still uses JSON.NET](https://github.com/aspnet/Configuration/blob/6c07ccd0a21952f68cfa5a7312797836bfdf51a2/src/Microsoft.Framework.ConfigurationModel.Json/project.json#L5). So, you can have comments there :smile:  enjoy the comments :trollface:

**Comment by tugberkugurlu** [comment]
> JSON is ostensibly a serialization format. This begs the question, why does a serialization format need comments?

:+1: to @adamralph.

OTOH, if we really care about comments at all on these files, let's not discuss forcing JSON to have comments. Instead, try to force the move to some other format like [YAML](http://yaml.org/) which actually [has comments](http://yaml.org/spec/1.2/spec.html#id2766934) and more suitable for hand written config files.

**Comment by Matthew-Bonner** [comment]
Other projects, such as Symfony allow you to use which you prefer, so rather than locking us down to JSON why not give us the choice of JSON, XML, TOML, JSON5, INI, etc? If other frameworks can do this surely ASP.NET can too. At least then for those who want comments they can at least use JSON5.

**Comment by adamralph** [comment]
@tugberkugurlu @Matthew-Bonner I like the direction you are steering in.

**Comment by davidfowl** [comment]
Even though at this point we're bike shedding and complaining I'll interject:

- We're using JSON, not any of those formats for the core project file. It's JSON. Configuration we don't care about as much, that can be in whatever format you choose, and that's why there's a provider model.
- Build systems and project systems don't let users choose what the file format is. Don't mix up configuration files with project files. They are different beasts.

**Comment by adamralph** [comment]
I don't think discussing a proposal to add comments to the JSON spec is bike shedding.

**Comment by davidfowl** [comment]
We must be reading different comments then.

**Comment by Matthew-Bonner** [comment]
Even though the comment at the top states, "the new JSON parser is customized for internal use. It is used to parse only the project.json, project.lock.json and global.json files." at some point I presume the Newtonsoft.Json JSON parser will be replaced in its entireity by the Microsoft.Framework.Runtime JSON parser so i think raising these issues now is worthwhile so that you can plan ahead.

**Comment by m0sa** [comment]
`git blame` only _somewhat_ works on `.json` files, because there are no trailing commas, which forces you to touch unrelated lines, just to add that comma where there was previously none.

As soon as you have the ability to add custom hooks in the project file (`scripts` and `commands` in project.json), you absolutely need comments. Take .csproj for example, there's usually no need for them, but if you start adding custom hooks (e.g. AfterBuild steps), you better add comments _because the guy who ends up maintaining the file will be a violent psychopath who knows where you live_. `git blame` usually turns out to be "fixing build" for those lines.

One alternative for project.json could be the ability to specify a `{ "command": "....", "description": "..."}` for `commands` and `scripts` instead of strings (and retrieve them via `dnx . web /?`)

**Comment by davidfowl** [comment]
@Matthew-Bonner You presume wrong.

**Comment by tugberkugurlu** [comment]
@davidfowl I generally agree on [this](https://github.com/aspnet/Announcements/issues/24#issuecomment-101570236) and personally, I don't care which format will be used for those runtime config files as long as there is a rich tooling support on it (which we have with JSON. E.g. Intellisense for NuGet packages on several editors thanks to JSON schema).

As for supporting multiple formats on runtime config (as @Matthew-Bonner stated), it is tempting to think that but we should really resist the temptation here :smile: and think about the side effects. If we think widely, it will only cause pain in the long run. Sticking with one bad format is better than supporting multiple good formats in this case. For your app config, choose whatever you want.

My main point is: if there will be a discussion here for comments o these config files, it shouldn't be around adding comments to JSON spec.

**Comment by LordZoltan** [comment]
Adding comments to the JSON spec is a discussion that should take place where ever the JSON spec is dealt with.  The arguments for and against are long and varied, and whilst I bemoaned the lack of them when starting to use JSON in anger outside of messaging, I agree with them not being there.
So the issue is instead that some of us want to be able to:
1 - disable blocks in project.json, say, without having to look in a file diff.  Why? Because the disabling might only be temporary (for example, if a bleeding edge dependency might be causing one framework target not to build, but you're expecting that to be fixed.  Had this exact problem recently with xunit).
2 - to communicate to other team members why something has been declared in project.json in a certain way (e.g. fixing a dependency version to non-latest because of compatibility issues, or why certain folder has to be included in global.json).
Simply saying 'you shouldn't need to do this' is not good enough, IMHO.
I personally find the lack of documentation in many of these files (yes, I know there is schema, but that only deals with the types things you can put in them, not *why* it's been done certain way in a specific file) makes them considerably harder to adopt and work with.
In the same way as a complex, but non-documented, msbuild script can be very difficult to work with, be it authored by someone else, or myself.
On 13 May 2015 9:31 am, David Fowler <notifications@github.com> wrote:We must be reading different comments then.

**Comment by davidfowl** [comment]
@LordZoltan I fail to see what 1. and 2. have to do with this entire issue.

**Comment by LordZoltan** [comment]
@DavidFowl
I guess I must be misunderstanding then.
Simply that by removing the ability to have comments (albeit when you can rightly argue they shouldn't be supported in the first place), disabling parts of these runtime files or annotating them for the benefit of other developers now becomes more difficult.
However, a simple readme might cover some of it...
I'm not going to get too precious on this ('could have fooled me' I'm sure some are thinking!), since the data format doesn't support it, then we can't have it.
Is it worth changing it all just to satisfy some people's need to have blocks of insignificant text? Probably not.
Do I prefer JSON as a file format for all this over XML? Absolutely.  So I'm willing to accept that compromise if there's no nice way around it, which I don't think there is at the moment.
On 13 May 2015 10:21 am, David Fowler <notifications@github.com> wrote:@LordZoltan I fail to see what 1. and 2. have to do with this entire issue.

**Comment by glen-84** [comment]
I felt very strongly about having comments included in the JSON spec, to the point that I was going to try to contact the spec writers and even organize a petition if necessary (ya, I'm crazy). However, I was looking at this completely the wrong way. JSON is supposed to be a data serialization format, *not* a configuration format.

Can we all agree on the following?:

1. That these project and solution files are meant to be human readable and editable. They are not in some binary format that only allows editing via CLI tools. There are tools within IDEs like VS to do syntax highlighting, package auto-completion, property auto-completion, etc.

2. That many developers would like the ability to include comments in these files, to temporarily remove blocks, and to document certain properties (why was something added?, is it temporary?, why this package and not that package?, what does this command do?, etc.).

3. That JSON itself is a data serialization format only, and was not intended to be used for configuration.

4. That adding support for comments in these files will cause confusion, as they are not valid in other such files (bower.json, package.json, etc.), and not part of the specification, meaning that various other parsers and editors will not handle them correctly (GitHub included).

If you agree with the above, I see only two possible options:

1. Make use of an appropriate format for configuration files. JSON5 might be a good option, since it's very similar to JSON.

2. If it is too late to make such a change, then we will have to stay with JSON, and support for comments should *not* be added.

Option 2 would really be unfortunate, but there may be valid reasons why switching formats at this stage would not be possible. It's still beta software though, so such changes can be expected.

**Comment by davidfowl** [comment]
> Make use of an appropriate format for configuration files. JSON5 might be a good option, since it's very similar to JSON.

`project.json`, `project.lock.json` and `global.json` are not configuration files, they are part of the new project system. I can't stress that enough. I also can't stress that we are *NOT* moving away from JSON for those particular formats.

**Comment by glen-84** [comment]
@davidfowl,

I respectfully disagree. They are used to *configure* the project – its version, dependencies, etc.

However, if you are adamant about not moving away from JSON, then I would suggest not supporting comments in the future.

**Comment by NickCraver** [comment]
@davidfowl Perhaps you guys should move to posting in a forum/format that doesn't allow feedback then? This was a regression, which it seems you've already set in stone. No offense, but just start the conversation with that next time so none of us waste our time supporting features the entire community loved.

**Comment by davidfowl** [comment]
@glen-84 Sure just like the csproj is considered a configuration file. The same class of file as `app.config` and `web.config`, makes sense. "Configuration" isn't overloaded at all in that sense :stuck_out_tongue_closed_eyes: 

@NickCraver Github actually lets you lock issues but we're not using it because we want to see the reaction people have when we make changes. 

It's ok for people to start healthy discussions about things, this is an open source and community driven project and we listen to everyone on multiple channels (twitter, jabbr, github, stackoverflow). 

That said:

This particular announcement is about the changes we made to the parser for those 3 files (as the description clearly states). We knew it would cause a fuss that's why we posed it and why I pointed people to it. That said, we are *NOT* changing the JSON format for those files, we just aren't at this point. We've invested too much time into that format (brand new editor in VS and VS code, project.json intellisense for nuget packags etc) to change it. As for the comments? We'll make a decision based on the amount of feedback about people needing them. That's why we haven't locked the thread.

I want to move the discussion away from picking different file formats (I don't see json5 as a different file format btw) and more to discussing why comments are the critical piece.

**Comment by glen-84** [comment]
@davidfowl,

Those files are XML, which supports comments. Configure means to "set up", but the semantics are unimportant. The main point is that it is a human-readable and editable file, and in such cases it is always useful to be able to include comments.

That said, I understand that there has been significant investment in this format. However, as you mentioned, JSON5 doesn't change the file format much, so I wonder how much effort would be required to reshape the JSON parser into a JSON5 parser and to make some small changes to the JSON editor to support JSON5 documents.

Regardless, I think the biggest mistake would be to add support for comments without changing the file extension (as mentioned by others above).

**Comment by davidfowl** [comment]
Hah, so all of the files would be project.json5 :smile: insanity.

**Comment by davidfowl** [comment]
I'm going to throw something out there to see how insane it is: What if we read both formats?

- If you have a project.json5 we'll use that and you can add all the comments in the world.
- If you have a project.json, it's strict JSON no comments allowed.

I'm not saying we'll do this btw but I want to see what people think (same for global.json but not project.lock.json, you don't edit that one).

**Comment by robertmuehsig** [comment]
:+1:

**Comment by Haplois** [comment]
:+1: Two file formats sounds great to me.

**Comment by glen-84** [comment]
@davidfowl,

Interesting idea, but I think that it may just complicate the project system, and make it less intuitive. Any external tool would then also need to handle both file formats. You would also have extra/conditional code, two parser dependencies, etc.

**Comment by csrakowski** [comment]
@davidfowl I for one am always up for choice and options, so I like that idea!

**Comment by NickCraver** [comment]
@davidfowl IMO that's a reasonable approach for the step forward in terms of opting in our out of compatibility with all a tool that doesn't support it. From a realistic standpoint though: whatever new projects are created with I would imagine to be the 99.5% use case with *just* that. 

*If* there was an easy way to switch in the tooling then it becomes tractable. For example, if you tried added a comment VS could prompt you to switch which would rename the file upon hitting Okay. If the tooling piece like that was in place, I think it's a much better story all around.

**Comment by davidfowl** [comment]
@glen-84 Sure but that's the nature of the beast. At least this way I can keep the `project.json` for the most part, but people strictly needing comments can get it working too (as long as there's enough JSON5 parsers in the world).

**Comment by davidfowl** [comment]
@NickCraver longer term maybe? Short term that wont happen.

**Comment by glen-84** [comment]
@davidfowl BTW, if you have project.json5, the lock file should be project.lock.json5, even if you strip out the comments. This way it can be collapsed in VS and also look consistent.

**Comment by davidfowl** [comment]
Sure

**Comment by SeanKilleen** [comment]
:thumbsup: this seems like a great idea. For most who don't need comments, they'd see no change. For those who do, a rename to `json5` allows comments. And, it sounds like it stays true to the specs of both from what I've read here. Yes, it could be slightly confusing but my guess is that most won't come into contact with it.

Though I have to say, given that, it seems to make sense to just default it to `.json5`. Allows comments from the outset, and anyone confused by the additional digit can google it. Remains true to its spec and allows things that folks might need. 

Informative discussion by the way; Thanks!

**Comment by Matthew-Bonner** [comment]
I guess we have moved on but still...
https://github.com/aspnet/Home/wiki/Project.json-file#configurations

Understandable that supporting JSON5 is something to be considered in respect of commenting, but VS reads and writes to these files, and I know I said earlier about supporting many formats but as you have clarified my comment is no longer relevant, so for the basis of the project.json if you have multiple formats does this not throw a few spanners in the works?

Consider the following, there is a projects.json and a projects.json5 file, which does VS prefer when it needs to make changes?

Also, despite making changes to the file myself, I can't merit the benefit of comments, and making the jump to support JSON5 is to me a bigger investment than some are considering, so is it really worth the time?

**Comment by PureKrome** [comment]
![image](https://cloud.githubusercontent.com/assets/899878/7609665/a3810e20-f9b8-11e4-8a43-b7ae01fef570.png)

**Comment by davidfowl** [comment]
@Matthew-Bonner nothing is set in stone, it was a straw man proposal to see if anyone though it was a good idea. We can't promise anything but we'll look at some options and see if anything is viable.

**Comment by nil4** [comment]
@davidfowl you seem to be open to the idea of supporting JSON5 file formats, which is great, and I want to add a :+1:. My understanding is that this change would imply updating dnx, nuget and VS tooling to be fully supported. Does this mean that the *ship has sailed* argument against renaming `project.json` (and perhaps `global.json`) to something more sensible in aspnet/dnx#1817 could still be reconsidered?

**Comment by vcsjones** [comment]
The more I think on the issue, the more I am willing to let go of comments in favor of following a specification. Two files to accomplish the same thing is, frankly, insane. Here we have a "clean slate" to do things right and simple, and this doesn't seem to align with that goal. If it's too late to change which specification we use, then I am happy to leave out comments. I can already see this conversation:

Q: Why is there a project.json and project.json5 option for my project?
A: The json5 file lets you use comments. Other than that, it is the same purpose.
Q: Oh. So I should use json5 for my config file then if I want comments?
A: Er, no. That uses json as well, but it allows comments.
Q: Oh, well, why doesn't the regular project.json file support comments like my config file does without this json5 stuff?

That just seems... awkward.

My approach to this is probably wrong. Maybe I think I need comments, when there is something better that I am missing, so let me explain why I think I need them, and perhaps someone can tell me why I am wrong.

I was quite looking forward to comments in project.json because it lets me add comments and history about the project in places that ASP.NET projects don't let you today. One particular example of that is depending on an old version of a nuget package. We Had a dependency on package, let's call it Frob, and used version 1.2, however there were newer versions on nuget. The new versions contained a subtle bug that was preventing us from upgrading until it was fixed.

In a project today, there is no really good place to visibly  capture that. The .csproj files are almost never looked at in their XML form except when merging, and are a poor place to annotate anything. Eventually someone would see in the Nuget package manager, "oh an update, I'll go ahead and update it" and then unit tests would break, and we'd have to communicate it - again. However with the .json file, we can just stick a comment next to our dependency: "//Do not upgrade beyond 1.2, newer versions have a bug we are waiting on a fix for."

We could have set up our own nuget feed that only has the exact versions of the nuget packages we want, but that "enterprise" approach felt heavy handed for what is just a minor problem.

Perhaps there is a better way to address the above without using comments.

**Comment by davidfowl** [comment]
@nil4 Nah we're not renaming it at this point.

**Comment by LordZoltan** [comment]
I'm going to be a total wimp here and backtrack on everything else I've said, although I guess that's what discussions are all about!
I'd like comments, but I'd rather not have to choose formats based on whether I want comments, not least because I fear for the IDE experience.  It's not that important to me.
So, in my very humble of opinions, if it's a choice between no comments or a choice of file formats, it'd be the former.
It's possibly the best solution to those for whom comments are a must, though - so kudos there.
On 13 May 2015 12:43 pm, David Fowler <notifications@github.com> wrote:@Matthew-Bonner nothing is set in stone, it was a straw man proposal to see if anyone though it was a good idea. We can't promise anything but we'll look at some options and see if anything is viable.

**Comment by tbertenshaw** [comment]
@davidfowl  What about if someone adds a project.json5 , upon save it runs the json5 converter to overright the project.json? Similar functionality to nested js and min.js files?

**Comment by mrtaikandi** [comment]
I don't get it. Before changing the parser there were comments and no mention of json5, two type of files, their extensions and their impact. 

So why not do the same? Why not just implement a "partial json5" parser and leave everything else as it is right now?

**Comment by glen-84** [comment]
> @davidfowl What about if someone adds a project.json5 , upon save it runs the json5 converter to overright the project.json? Similar functionality to nested js and min.js files?

Too much magic which would only work in VS, IMHO.

**Comment by glen-84** [comment]
> So why not do the same? Why not just implement a "partial json5" parser and leave everything else as it is right now?

Because you should be able to read a *.json file with any spec-compliant parser. If you add comments, you break existing tools and parsers.

In my opinion, the choice should be between project.json *without* comments and project.json5 (or similar) *with* comments.

**Comment by mariusschulz** [comment]
@glen-84 :thumbsup:

**Comment by glen-84** [comment]
What makes this even more tricky is that if a decision was made to use something like JSON5, people would also want to use it for their configuration files. In such case, there would have to be two implementations, in order to avoid the same potential version conflicts that motivated the development of this JSON parser in the first place.

For this reason, if comments are deemed necessary by the ASP.NET team (perhaps after gathering feedback), it may make more sense to create a simple custom project file format that just adds comment support to a JSON-like document, and uses a different file extension like *.dnx (f.e. project.dnx).

The issue then would be editor support, as the editor would need to understand three file types (json, json5, and dnx). Perhaps just supporting strict JSON and JSON5 would be more viable.

**Comment by rajajhansi** [comment]
I vote for a single format i.e., project.json5. Even thought I don't use *nix and OSX, as long as VS2015 and the command line tool dnvm that uses project.json5 parses it correctly and MS ships the json5 parser,IDE and command line tools, what is the issue? Btw, VS still uses .sln, a file that command line tools don't care about.

**Comment by sebastiang** [comment]
You should totally use XML! It has support for comments! It could look something like this

    <Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
        <Reference Include="System.Reactive.Linq" />
        <Compile Include="ApplicationTest.cs" />
        ...

**Comment by FransBouma** [comment]
Man, the progress made with these new formats. XML had comments since before the first world war, and was perfectly fine as a project format. Opting for 'json' is not making _more_ things possible, but _less_ things possible. I.o.w.: it's limiting. While there are so many other formats out there which don't force these limits onto you. 

I find it rather funny that there are people in this thread who spent a tremendous amount of energy to apologize for them being limited by this format. "It's a good thing!". Sure it is, in your world, but you're not everyone, and this _is_ for everyone.

**Comment by gregoryyoung** [comment]
We implemented a new JSON parser in Microsoft.Framework.Runtime and removed the dependency on Newtonsoft.Json.

"Using Newtonsoft.Json in the runtime prevents customers from choosing a specific version of Newtonsoft.Json in their applications. This change will allow customers to use any version of Newtonsoft.Json in their applications."

Why not just internalize the dependency in which case the problem would go away?

**Comment by markrendle** [comment]
Sod JSON. Use [MOFO](https://github.com/markrendle/mofo).

:wave:
:microphone:

**Comment by glen-84** [comment]
@gregoryyoung,

    darkangel: That would be fixed by internal deps one day maybe.     1:54:01 PM
    davidfowl: yea if you want to internalize 900 KB                   1:54:22 PM
    davidfowl: or however big json.net is                              1:54:27 PM
    davidfowl: when all you need a super simple JSON DOM               1:54:37 PM
    darkangel: sure                                                    1:54:44 PM
    darkangel: Trade offs.                                             1:54:49 PM

**Comment by daveaglick** [comment]
I'm with @FransBouma. I understand the desire to get away from XML, like just about everyone else I'm not a huge fan of it's verbose syntax. For that matter, I'm not a huge fan of the quotes-everywhere nature of JSON either. But ending up with a solution that has fewer features than what we're leaving behind in the name of syntax seems...odd. I don't know what the best language would be, but I'm guessing it's a moot point because the JSON vs. YAML vs. HipsterMarkupOfTheMonth discussions happened a while ago?

That said, if JSON is "it" I would advocate for some way to keep the commenting ability. Remember that this is .NET and there is a huge enterprise userbase out there that will be adopting these new technologies with all kinds of different use cases, conventions, etc. Most aren't on GitHub, Jabbr, etc. but I suspect at least some currently use and would like to continue using comments in their various configuration and project artifacts.

**Comment by ironyx** [comment]
I agree fully with @daveaglick - but I must say that this discussion is the greatest example of Parkinson's Law of Triviality I've seen in a long time. Does it *really* matter that much in the grand scheme of vNext changes? I feel we should all consider the importance of this change relative to everything else.

**Comment by Romoku** [comment]
JSON is a bunch of name-value pairs. As long as the user can understand what the keys and values mean then there is no need for comments.

**Comment by uatec** [comment]
I disagree. If i have a list of nuget references, I want to group them and
say "x, y and z are all related to thing M"...

**Comment by luisrudge** [comment]
:-1: For multiple file formats.

**Comment by skoon** [comment]
Who gives a crap, DEATH TO ALL PROJECT FILES!!!!

**Comment by GuardRex** [comment]
:+1: for [JSON5](http://json5.org/) as an optional format or the main format. I'll use it with comments.

For those of us in favor, please put votes in at Stephen Cleary's [Use JSON5 for project.json at Uservoice](https://aspnet.uservoice.com/forums/252111-asp-net-vnext/suggestions/7209422-use-json5-for-project-json).

**Comment by GuardRex** [comment]
Shoutout to Aseem Kishore @aseemk ... at [JSON5 on Github](https://github.com/aseemk/json5).

**Comment by sblackler** [comment]
Personally, have two different sets of files will be confusing. I think it needs to be one way or the other. I don't think they do need comments in the 3 files mentioned, but not objective to having them also.

**Comment by MFFoX** [comment]
I have to completely agree with @NickCraver's first comment, specifically

> Comments are needed here, and in many, many other similar scenarios. This is an excellent use case to actually advance the JSON spec.

 IMHO, it is not your responsibility to keep us honest by enforcing the spec. If other tools don't support comments, then it is our responsibility to go and fight for another implementation, conform our `.json` to the spec, or stop relying on that tool. 

In my eyes, the only reason that would be worth conforming to the spec is because you rely on that functionality. This is not the case. Conforming to a standard just for the sake of conformity is not worth it. 

Finally, keep in mind what you're implying with this change. We're replacing xml `.config` files with `project.json` files. How many configs have you worked with that have zero comments? The suggestions for using specific git features in place is asking _way_ more of us than understanding that other tools might not understand json with comments. 

TL;DR my +1 is for comments support and breaking the spec for usability.

**Comment by rossipedia** [comment]
The responses suggesting using git are missing the boat entirely. This isn't about using comments to disable or remove part of the file temporarily, this is about using comments as they are intended: _commenting_ on a part of the file (to add clarity of intent).

I understand why Crockford removed them, I understand they're not part of the spec, but in this case, both are just plain _wrong_.

**Comment by glen-84** [comment]
> TL;DR my +1 is for comments support and breaking the spec for usability.

According to [Douglas Crockford](https://groups.yahoo.com/neo/groups/json/conversations/topics/152):

> JSON does not have comments. A JSON encoder MUST NOT output comments.
> A JSON decoder MAY accept and ignore comments.

> Comments should never be used to transmit anything meaningful. That is
> what JSON is for.

In other words (if this idea was included in the spec), it is *valid* for a decoder to accept and ignore comments.

HOWEVER, in my opinion, it causes unnecessary confusion – why can I comment in project.json but not bower.json or package.json? Also, it breaks in editors and tools that do not support comments. JSON5 also adds a number of other really useful features.

**Comment by MFFoX** [comment]
> Comments should never be used to transmit anything meaningful. That is
what JSON is for.

In that case, this is a no brainer. Thanks for including that bit :)

 I don't really agree with you about it being confusing. It isn't confusing to me when I try to run a tool that doesn't support comments in json and it errors with an exception to that extent. That's fairly straightforward.

**Comment by Eilon** [comment]
Hi everyone,

Wow, what a discussion! We've had similar discussions within the team about supporting comments in the `project.json`, `project.lock.json`, and `global.json` files. We've also discussed whether we should switch to different file formats as an alternative.

Ultimately, we decided that *for now* it is best to stick with the industry standard JSON format, which has a [very well-defined specification](http://json.org/). We have chosen JSON for the following reasons:

1. Well, it's an industry standard. That's pretty neat.
2. It has support in nearly every tool, web site, editor, parser, which leads to great interoperability.
3. It is an easy-to-understand, human-readable/writeable, computer-readable/writeable format.
4. It supports the right set of data structures: arrays, dictionaries, strings, numbers, Booleans.

Though we're not opposed to changing this in the future, this is where we are right now. The purpose of this issue is to announce the change and to see if people had additional thoughts on the effect or consequences of the change. We've carefully read the entire discussion here and will still look at changing this in the future.

I'm now locking this thread because I think the purpose of this thread has been served.

If you have further thoughts on this you can reach me at elipton-AT-microsoft.com.

Thanks,
Eilon Lipton (Engineering Manager for ASP.NET and EntityFramework)
